protobuf/lib/src/protobuf/mixins/well_known.dart

```diff
@@ -65,4 +162,573 @@ abstract class TimestampMixin {
+      throw ArgumentError(
+          'Timestamp with `nanos` out of range: ${timestamp.nanos}');
+    }
+    if (dateTime.isBefore(_minTimestamp) || dateTime.isAfter(_maxTimestamp)) {
+      throw ArgumentError('Timestamp Must be from 0001-01-01T00:00:00Z to '
+          '9999-12-31T23:59:59Z inclusive. Was: ${dateTime.toIso8601String()}');
+    }
+
+    // Because [DateTime] doesn't have nano-second precision, we cannot use
+    // dateTime.toIso8601String().
+    String y = '${dateTime.year}'.padLeft(4, '0');
+    String m = _twoDigits(dateTime.month);
+    String d = _twoDigits(dateTime.day);
+    String h = _twoDigits(dateTime.hour);
+    String min = _twoDigits(dateTime.minute);
+    String sec = _twoDigits(dateTime.second);
+    String secFrac = timestamp.nanos
+        .toString()
+        .padLeft(9, '0')
+        .replaceFirst(finalGroupsOfThreeZeroes, '');
+    if (secFrac != '') secFrac = '.$secFrac';
+    return "$y-$m-${d}T$h:$min:$sec${secFrac}Z";
+  }
+
+  static void fromProto3JsonHelper(GeneratedMessage message, Object json,
+      TypeRegistry typeRegistry, JsonParsingContext context) {
+    if (json is String) {
+      Match fracSecsMatch = RegExp(r'\.(\d*)').firstMatch(json);
+      String fracSecs = fracSecsMatch?.group(1) ?? '0';
+      if (fracSecs.length > 9) {
+        throw context
+            .parseException('Timestamp with more than 9 decimal digits '
+                'does not fit into nanoseconds');
+      }
+      int nanos = int.tryParse(fracSecs.padRight(9, '0')) ?? 0;
+
+      String jsonWithoutFracSec = fracSecsMatch == null
+          ? json
+          : json.replaceRange(fracSecsMatch.start, fracSecsMatch.end, '');
+      DateTime dateTimeWithoutFractionalSeconds;
+      try {
+        dateTimeWithoutFractionalSeconds = DateTime.parse(jsonWithoutFracSec);
+      } on FormatException {
+        throw context.parseException('Timestamp not well formatted: $json');
+      }
+
+      TimestampMixin timestamp = message as TimestampMixin;
+      setFromDateTime(timestamp, dateTimeWithoutFractionalSeconds);
+      timestamp.nanos = nanos;
+    } else {
+      throw context.parseException(
+          'Expected timestamp represented as String, found: $json');
+    }
+  }
+}
+
+abstract class DurationMixin {
+  Int64 get seconds;
+  set seconds(Int64 value);
+
+  int get nanos;
+  set nanos(int value);
+
+  static final RegExp finalZeroes = RegExp('(0)*\$');
+
+  static Object toProto3JsonHelper(
+      GeneratedMessage message, TypeRegistry typeRegistry) {
+    DurationMixin duration = message as DurationMixin;
+    String secFrac = duration.nanos
+        // nanos and seconds should always have the same sign.
+        .abs()
+        .toString()
+        .padLeft(9, '0')
+        .replaceFirst(finalZeroes, '');
+    String secPart = secFrac == '' ? '' : '.$secFrac';
+    return "${duration.seconds}${secPart}s";
+  }
+
+  static final RegExp durationPattern = RegExp(r'(-?\d*)(?:\.(\d*))?s$');
```

Comment:
I can see you allow and handle `.5s` and `5.s` and `.s`.
Is this deliberate? (I doesn't seem like *valid* inputs, especially '.s').

Reply:
The spec says 'any number of digits'. I guess any includes 0.